src/core/reducer.ts

```diff
@@ -121,6 +128,77 @@ function initializeDeltalog(
   };
 }
 
+/**
+ * ExtractTransientsFromState
+ *
+ * Split out transients from the a TransientState
+ */
+function ExtractTransients(
+  transientState: TransientState
+): [State | null, TransientMetadata | null] {
+  if (!transientState) {
+    // We preserve null for the state for legacy callers, but the transient
+    // field should be undefined if not present to be consistent with the
+    // code path below.
+    return [null, undefined];
+  }
+  const { transients, ...state } = transientState;
+  return [state as State, transients as TransientMetadata];
+}
+
+/**
+ * WithError
+ *
+ * Augment a State instance with transient error information.
+ */
+function WithError<PT extends any = any>(
+  state: State,
+  errorType: ErrorType,
+  payload?: PT
+): TransientState {
+  const error = {
+    type: errorType,
+    payload,
+  };
+  return {
+    ...state,
+    transients: {
+      error,
+    },
+  };
+}
+
+/**
+ * Middleware for processing TransientState associated with the reducer
+ * returned by CreateGameReducer.
+ * This should pretty much be used everywhere you want realistic state
+ * transitions and error handling.
+ */
+export const TransientHandlingMiddleware = (store: Store) => (
+  next: Dispatch<ActionShape.Any>
+) => (action: ActionShape.Any) => {
+  const rawResult = next(action);
+  switch (action.type) {
+    case Actions.STRIP_TRANSIENTS: {
+      return rawResult;
+    }
+    default: {
+      const [, transients] = ExtractTransients(store.getState());
+      if (typeof transients !== 'undefined') {
+        const stripTransientsResult = store.dispatch(stripTransients());
+        // TODO(#723): Revisit this result payload and consider converting
+        // to a promise.
+        return {
+          ...rawResult,
+          stripTransientsResult,
```

Comment:
Do we need to include `stripTransientsResult` here? What might it be used for?

Reply:
I don't have a concrete use case yet, and it seems like this detail isn't very important, but here's my long-winded thinking behind this decision. While not necessarily pertinent to this change, it may provide an opportunity for you to tell me I'm thinking about this totally wrong and save a lot of time going forward:

 1. I'm basically thinking about this like a redux-friendly exception stack pattern where each middleware acts as a stack frame. (I just haven't implemented a stack yet, because there's only one virtual-action-spawning middleware for now, and creating a more general API for adding and consuming from a stack with only one data point probably be a waste of time.)

 2. Other middleware used alongside this one should generally make a decision about whether to specifically handle "virtual" actions (like the auto-generated STRIP_TRANSIENTS) or just pass them through untouched with a case statement on the action type. For these middlewares that don't care, it would be easier for them to delegate the multiplexing of the result values to the the middleware that spawns the "virtual" action (i.e. this one).

 3. For middlewares that might actually care (like loggers or bots), it might be useful, for visualization/debug/tree-pruning purposes, to know the provenance of the virtual action (i.e. which "real" action caused the virtual action to be spawned). In the current setup, there's only a single virtual action so such a logger middleware could simply maintain some state and look back at the previous action received, but this is messy and won't generalize if we expand on this "virtual" action pattern.
    Imagine if we introduce a second action-spawning middleware in our stack (e.g. to automatically spawn master->client synchronization actions that might allow us to subsequently unify the client and master logic).
   Then (absent some stack-based result-multiplexing) the implementation of a provenance-tracking logger would have to depend on the exact stack-order of the virtual-action-spawning middlewares... plausible, but nasty!

 4. Because the spawned "virtual" action is also exposed to the other middle-wares such that it could be modified/filtered on its way to and from the reducer, it is useful to know what the actual result value is at the handleTransients layer. If another middleware is added before this one (as we've discussed), 

 5. As for why I'm adding the "virtual" result relation to the result of the "real" action and not the other way around (i.e. adding the spawning action to the virtual action result), this is just because of the stack-resolution order of the actions. While the incoming dispatch order is real action -> virtual action, the dispatch return order is virtual action -> real action. Thus, if something fails in the middleware handling of the virtual action, useful forensic information could be lost.

Reply:
Hmm, OK. To the extent that I understand this, I still think this could be a case for [“YAGNI”](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). For now, the action shape will be `{ type, payload }` or `{ type, payload } & { transients, stripTransientsResult }`, i.e. presence of `transients` indicates transients were stripped and the actual value of `stripTransientsResult` is always the same so it doesn’t actually carry any significant information. Another middleware could achieve all the same functionality with just `{ type, payload, transients? }`.

Happy for you to explore stacks etc. at some point, but for now I’d keep this as simple as possible for what we need. Augmenting the situation later should be a simple change.

Reply:
![image](https://user-images.githubusercontent.com/129236/119027912-dd97f380-b974-11eb-8745-734d7ca52ac2.png)

Reply:
Cool! I guess all I’m saying is let’s avoid adding data that is not actually consumed yet. If a future PR needs the `stripTransientsResult` property, let’s add it then.

Reply:
Removed the payload, updated the master.test, and added a Dev Note with a link to this thread, since it's a bit trickier than it looks and I think this discussion/diagram would be useful to anybody tempted to muck with this code in the future..